### ISSUE TITLE
base_state problems

### ISSUE THREAD

**Issue by AntonReinhard** [QEDjl-project/QEDbase.jl#36]
There are two things:

First, the `base_state` doesn't behave as it should with `AllPol` and `AllSpin`:
```julia
julia> base_state(Photon(), Incoming(), SFourMomentum(1.0, 0., 0., 0.), AllPol())
ERROR: MethodError: no method matching _photon_state(::AllPolarization, ::SFourMomentum)

Closest candidates are:
  _photon_state(::PolarizationX, ::QEDbase.AbstractFourMomentum)
   @ QEDbase ~/.julia/packages/QEDbase/wLSbk/src/particles/particle_states.jl:186
  _photon_state(::PolarizationY, ::QEDbase.AbstractFourMomentum)
   @ QEDbase ~/.julia/packages/QEDbase/wLSbk/src/particles/particle_states.jl:194

Stacktrace:
 [1] base_state(particle::Photon, #unused#::Incoming, mom::SFourMomentum, pol::AllPolarization)
   @ QEDbase ~/.julia/packages/QEDbase/wLSbk/src/particles/particle_states.jl:212
 [2] top-level scope
   @ REPL[35]:1

julia> base_state(Electron(), Incoming(), SFourMomentum(1.0, 0., 0., 0.), AllSpin())
ERROR: MethodError: no method matching _spin_index(::AllSpin)

Closest candidates are:
  _spin_index(::SpinUp)
   @ QEDbase ~/.julia/packages/QEDbase/wLSbk/src/particles/particle_spin_pol.jl:51
  _spin_index(::SpinDown)
   @ QEDbase ~/.julia/packages/QEDbase/wLSbk/src/particles/particle_spin_pol.jl:52
  _spin_index(::AbstractDefiniteSpin)
   @ QEDbase ~/.julia/packages/QEDbase/wLSbk/src/particles/particle_spin_pol.jl:50

Stacktrace:
 [1] base_state(particle::Electron, #unused#::Incoming, mom::SFourMomentum, spin::AllSpin)
   @ QEDbase ~/.julia/packages/QEDbase/wLSbk/src/particles/particle_states.jl:125
 [2] top-level scope
   @ REPL[36]:1
```

For photons, the function `function _photon_state(mom::QEDbase.AbstractFourMomentum)` exists but has a different signature, so it is not dispatched to.

Maybe these examples could also be used as doc test strings (after this is fixed), since they are one-liners.

Second, the return type is only a `Vector{...}` when an indefinite Spin/Polarization is passed, which makes sense but also defeats the purpose for the user since if they try to iterate over the `base_state`'s return value, they will iterate through the values of the `Spinor`/`LorentzVector`, if only a single one was returned. There also seems no simple way to get the result into a single "layer" of vector, regardless of the original return type (vector/non-vector).

I'd suggest always returning a Vector, even if only one value is returned.

**Comment by szabo137** [comment]
Thanks for opening this issue, the behavior is indeed not correct, or at least misleading. 

Regarding the suggested solution, I am not a fan of returning one-element vectors if the user might expect the element, but I understand the iterator argument. For the latter purpose, I suggest a simple `as_vector` function, which wraps the result of a single `base_state` in a one-element vector but does nothing for inputs, which are already vectors. For instance:

```julia
@inline _as_svec(x) = SVector((x,))
@inline _as_svec(x::SVector{N,T}) where {N,T<:AbstractState} = x
@inline _as_svec(x::NTuple) = SVector(x)
``` 
(the latest is just convenience, `SVector` is just for speed). 

The iterator call could then look like

```julia
iters = Iterators.product(_as_svec.([el_states,ph_states,el2_states]))
```
where the states could be singles or vectors (or tuples).

**Comment by AntonReinhard** [comment]
Solved in #37
